docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+so Desktop and iOS can fit into this world.
+
+This state management should also provide the correct degree of isolation for
+the various components. For example, each engine should only know about state
+which directly impacts how it works. For example, the keys used to encrypt
+a collection should only be exposed to that specific engine, and there's no
+need for one engine to know what info/collections returns for other engines,
+nor whether the device is currently connected to WiFi.
+
+A thorn here is for persisted state - it would be ideal if the low-level
+component could avoid needing to persist any state, so it can avoid any
+kind of storage abstraction. A solution here might be that the low-level
+component can delegate state storage to the high-level component in an
+opaque way.
+
+# Implementation plan for the low-level component.
+
+Let's try and move into actionable decisions for the implementation. We expect
+the implementation of the low-level component to happen first, followed very
+closely by the implementation of the high-level component for Android. So we
+focus first on these.
+
+## Clients Engine
+
+The clients engine includes some meta-data about each client. We've decided
+we can't replace the clients engine with the FxA device record and we can't
+simply drop this engine entirely.
+
+Of particular interest is "commands" - these involve communicating with the
+engine regarding commands targetting it, and accepting commands to be send to
+other devices. Note that outgoing commands are likely to not originate from a sync,
+but instead from other actions, such as "restore bookmarks".
+
+However, because the only current requirement for commands is to wipe the
+store, and because you could anticipate "wipe" also being used when remotely
+disconnecting a device (eg, when a device is lost or stolen), our lives would
+probably be made much simpler by initially supporting only per-engine wipe
+commands.
+
+## Collections vs engines vs stores vs preferences
+
+For the purposes of the sync manager, we define:
+
+* An *engine* is the unit exposed to the user - an "engine" can be enabled
+  or disabled. There is a single set of canonical "engines" used across the
+  entire sync ecosystem - ie, desktop and mobile devices all need to agree
+  about what engines exist and what the identifier for an engine is.
+
+* A *store* is the code which actually syncs. This is largely an implementation
+  detail. There may be multiple stores per engine (eg, the "history" engine
+  may have "history" and "forms" stores.)
+
+* A *collection* is a unit of storage on a server. It's even more of an
+  implementation detail than a store. For example, you might imagine a future
+  where the "history" store uses multiple "collections" to help with containers.
+
+In practice, this means that the high-level component should only need to care
+about an *engine*. The low-level component will manage the mapping of engines
+to stores.
```

Comment:
The app will independently need to know about stores though right, in order to actually use the store for providing application-level functionality?

Is it ever the case that a single store corresponds to multiple engines?  I expect yes, with the "places store" corresponding to both the "history" and "bookmarks" engines.

I wonder if there are any opportunities to clean up our historical baggage here.  For example, if we implemented future forms functionality as part of the places store, would that avoid the need to allow for multiple stores per engine?

Reply:
> right, in order to actually use the store for providing application-level functionality?

Yes, this section is bad - naming is hard. It also gets to the root of the "sling instances around" comment above.

The best name, for the thing that provides application-level functionality, and given their names in the Rust code, is probably an "Api". The stores I refer to here are the [lower-level sync stores](https://github.com/mozilla/application-services/blob/master/components/sync15/src/sync.rs#L15). The "PlacesApi" has 2 sync-stores - bookmarks and history. The logins api has 1 sync store.

The sync manager can hopefully be done without knowing about any Apis and just about stores.

The framing of this part of the doc needs more work - it's from the inside looking out - thanks!